src/snowflake/snowpark/mock/_plan.py

```diff
@@ -443,32 +454,56 @@ def cleanup_imports():
                 if module_path in sys.path:
                     sys.path.remove(module_path)
 
-        # Clear sys.modules cache
+        # Clear added entries in sys.modules cache
         added_keys = set(sys.modules.keys()) - frozen_sys_module_keys
         for key in added_keys:
             del sys.modules[key]
 
+        # Cleanup import directory
+        analyzer.session.udf._udf_import_directories[udf_name].cleanup()
+
+        # Restore snowflake_import_directory
+        if last_import_directory is not None:
+            sys._xoptions["snowflake_import_directory"] = last_import_directory
+        else:
+            del sys._xoptions["snowflake_import_directory"]
```

Comment:
what is the `Restore` process here?

Reply:
We need `last_import_directory` when a UDF's input is another UDF's results. On line 510, we calculate the child expression as function input, when this involves another UDF expression, the child UDF should be computed with the corresponding import directory but after its execution this should be restored to the last UDF execution on the stack, so we need to save `last_import_directory` in each `handle_udf_expression` call and restore it at the end.

Reply:
I see, that makes more sense to me now, thanks for the explanation.